ocean_lib/ocean/v4/ocean_assets.py

```diff
@@ -0,0 +1,200 @@
+        nft_address: str,
+        files: str,
+        provider_uri: Optional[str] = None,
+        nft_name: Optional[str] = None,
+        nft_symbol: Optional[str] = None,
+        template_index: Optional[int] = 1,
+        nft_additional_erc_deployer: Optional[str] = None,
+        nft_uri: Optional[str] = None,
+    ) -> Optional[V4Asset]:
+        assert isinstance(
+            metadata, dict
+        ), f"Expected metadata of type dict, got {type(metadata)}"
+
+        # copy metadata so we don't change the original
+        metadata_copy = copy.deepcopy(metadata)
+
+        asset_type = metadata_copy["type"]
+        assert asset_type in (
+            "dataset",
+            "algorithm",
+        ), f"Invalid/unsupported asset type {asset_type}"
+
+        validation_result, validation_errors = self.validate(metadata)
+        if not validation_result:
+            msg = f"Metadata has validation errors: {validation_errors}"
+            logger.error(msg)
+            raise ValueError(msg)
+
+        if not provider_uri:
+            provider_uri = DataServiceProvider.get_url(self._config)
+
+        services = services or []
+        services = self._add_defaults(services, nft_address, files, provider_uri)
+        # Create a DDO object
+        asset = V4Asset()
+
+        address = get_address_of_type(self._config, ERC721FactoryContract.CONTRACT_NAME)
+        erc721_factory = ERC721FactoryContract(self._web3, address)
+
+        if not nft_address:
+            name = nft_name or metadata["name"]
+            symbol = nft_symbol or name
+            additional_erc20_deployer = nft_additional_erc_deployer or ZERO_ADDRESS
+            token_uri = nft_uri or "https://oceanprotocol.com/nft/"
+            # register on-chain
+            tx_id = erc721_factory.deploy_erc721_contract(
+                name,
+                symbol,
+                template_index=template_index,
+                additional_erc20_deployer=additional_erc20_deployer,
+                token_uri=token_uri,
+                from_wallet=publisher_wallet,
+            )
+            _ = self._web3.eth.wait_for_transaction_receipt(tx_id)
+            nft = ERC721Token(self._web3, erc721_factory.get_token_address(tx_id))
+            if not nft:
+                logger.warning("Creating new data token failed.")
+                return None
+            logger.info(
+                f"Successfully created data token with address "
+                f"{nft.address} for new dataset asset."
+            )
+        else:
+            # verify nft address
+            if not erc721_factory.verify_nft(nft_address):
+                raise ContractNotFound(
+                    f"NFT address {nft_address} is not found in the ERC721Factory events."
+                )
+
+        assert nft_address, "nft_address is required for publishing a dataset asset."
+
+        # Generating the did and adding to the ddo.
+        did = f"did:op:{remove_0x_prefix(nft_address)}"
```

Comment:
did has a new format, see in the docs, and implemented in the provider and aquarius tests.